### DISCUSSION TITLE
Unable to Run pip After Installing Python via mise on Windows 11

### DISCUSSION THREAD

**Discussion by gjlmotea** [jdx/mise#3821] [Troubleshooting and bug reports]
**Describe the bug**
Hello!
I am new to mise and really love its elegant and intuitive operation. Thank you for providing such a wonderful tool!

I don't know why I can't run `pip` after install python.

**To Reproduce**
Windows11 powershell
```
winget install jdx.mise

$shimPath = "$env:USERPROFILE\AppData\Local\mise\shims"
$currentPath = [Environment]::GetEnvironmentVariable('Path', 'User')
$newPath = $currentPath + ";" + $shimPath
[Environment]::SetEnvironmentVariable('Path', $newPath, 'User')

mise use --global python@3.12
```

Additionally, I had to follow this guide to disable Microsoft Store's redirection when typing python
https://stackoverflow.com/questions/58754860/cmd-opens-windows-store-when-i-type-python

Finally, I got this:
```
> python
python 3.12.8

> python3
(not found)

> python3.12
(not found)

> pip
(not found)

> python -m pip install --upgrade pip
Requirement already satisfied: pip in c:\users\owner\appdata\local\mise\installs\python\3.12.8\lib\site-packages (24.3.1)
```

**Expected behavior**
python3, python3.12, pip command is available

**`mise doctor` output**

```
PS C:\Users\OWNER> mise doctor
version: 2024.12.6 windows-x64 (7c736f2 2024-12-11)
shims_on_path: yes

build_info:
  Target: x86_64-pc-windows-msvc
  Features: DEFAULT, NATIVE_TLS, OPENSSL
  Built: Wed, 11 Dec 2024 17:30:27 +0000
  Rust Version: rustc 1.83.0 (90b35a623 2024-11-26)
  Profile: release

shell:
  (unknown)

dirs:
  cache: ~\AppData\Local\Temp\mise
  config: ~\.config\mise
  data: ~\AppData\Local\mise
  shims: ~\AppData\Local\mise\shims
  state: ~\.local\state\mise

config_files:
  ~\.config\mise\config.toml

backends:
  aqua
  asdf
  cargo
  core
  gem
  go
  npm
  pipx
  spm
  ubi
  vfox

plugins:

toolset:
  core:python@3.12.8

env_vars:
  (none)

settings:
mise WARN  mise version 2024.12.19 available
mise WARN  To update, run mise self-update
```

**Comment by ssbarnea** [comment]
While my use of Windows is very limited, I faced this issue while trying to test mise-action on all platforms. Apparently the current instructions for installing on Windows are not really working.

Scoop can be ignored for the moment as the PR was not merge yet and winget command should look like:

```powershell
winget install --accept-source-agreements jdx.mise
# really needed because the store published version is always outdated
mise self-update -y
# needed to avoid python AppStore install prompts from Windows
del /f /q "%USERPROFILE\AppData\Local\Microsoft\WindowsApps\python*.exe" >nul 2>&1
```

```shell
# later after manually updating PATH to include shims and a system reboot:
C:\Users\ssbarnea>echo %PATH%
C:\WINDOWS\system32;C:\WINDOWS;C:\WINDOWS\System32\Wbem;C:\WINDOWS\System32\WindowsPowerShell\v1.0\;C:\WINDOWS\System32\OpenSSH\;C:\Program Files\PowerShell\7\;C:\Users\ssbarnea\AppData\Local\Microsoft\WindowsApps;C:\Users\ssbarnea\AppData\Local\Microsoft\WinGet\Links;C:\Users\ssbarnea\AppData\Local\mise\shims;

C:\Users\ssbarnea>type python
The system cannot find the file specified.

C:\Users\ssbarnea>dir C:\Users\ssbarnea\AppData\Local\mise\shims

 Directory of C:\Users\ssbarnea\AppData\Local\mise\shims

30/12/2024  10:44    <DIR>          .
30/12/2024  10:44    <DIR>          ..
30/12/2024  10:44                39 python.cmd
30/12/2024  10:44                40 pythonw.cmd
30/12/2024  10:40                37 task.cmd
```

Sadly even after these, mise seems to be broken:
```
PS C:\Users\ssbarnea> python
mise ERROR cannot find binary path
mise ERROR Run with --verbose or MISE_VERBOSE=1 for more information

For more information, try '--help'.
PS C:\Users\ssbarnea> mise --verbose exec -- python
DEBUG ARGS: C:\Users\ssbarnea\AppData\Local\Microsoft\WinGet\Links\mise.exe --verbose exec -- python
DEBUG EnvResults
DEBUG EnvResults
DEBUG EnvResults
Error:
   0: cannot find binary path

Location:
   src\cli\exec.rs:138

Version:
   2024.12.22 windows-x64 (9d613b9 2024-12-30)

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
```

## Issues found so far
- [ ] incomplete winget command
- [ ] outdated version being installed (including self-update after should be included in install)
- [ ] `type -a mise` fail with `The system cannot find the file specified.` but running mise is somehow  works
- [ ] PATH not updated by default and mise is not functional without, so installation is broken. Install snippet should be *complete*, so after running it mise would be fully functional.
- [ ] Even after I manually added shims to PATH, shell is not able to find commands installed there

**Comment by bharat-mon** [comment]
I've same issues with winget install and python (specifically). Java and node are working fine. Installing mise via scoop is working for python without any issues.

The problem with pip specifically is unlike the node/npm, there is no shim created for pip. Here is the output I got while manually upgrading pip:

```
> python -m pip install --upgrade pip

Requirement already satisfied: pip in c:\users\bharat\scoop\apps\mise\current\mise\installs\python\3.13.1\lib\site-packages (24.3.1)
Collecting pip
  Downloading pip-25.0-py3-none-any.whl.metadata (3.7 kB)
Downloading pip-25.0-py3-none-any.whl (1.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 22.2 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 24.3.1
    Uninstalling pip-24.3.1:
      Successfully uninstalled pip-24.3.1
  WARNING: The scripts pip.exe, pip3.13.exe and pip3.exe are installed in 'C:\Users\Bharat\scoop\apps\mise\current\mise\installs\python\3.13.1\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed pip-25.0
```

So manually adding the Scripts folder to path fixed my issue with pip.

**Comment by crazydan** [comment]
Similar issue here. It seems like mise isn't putting the Scripts folder for the current Python version into the PATH env var. It also doesn't create the python3, pip3 aliases for the Python distribs.

**Comment by xvrpop** [comment]
The issue persists, I already had to add manually the shims directories to the PATH (I think mise activate should be simplified). Python is installed correctly, but I can't use call pip directly because no shims have been created for it.  Seems like the issue may be caused by the [quirks of the precompiled binaries](https://github.com/astral-sh/python-build-standalone/blob/main/docs/quirks.rst#no-pipexe-on-windows).

**Comment by SuSuSoo** [comment]
I'm experiencing the same issue. After installing Python 3 globally with mise, I cannot use pip globally, but this problem doesn't occur when using pyenv.

System Information:
- OS: Microsoft Windows 10 Enterprise 2021 LTSC
- Shell: PowerShell 7.5+
- mise version: 2025.11.11 windows-x64 (2025-11-30)

After analysis, I believe the reason is that mise doesn't create corresponding shims or environment variables for pip. On Windows, pip.exe is typically installed in the Scripts directory, but mise seems to only add the Python main executable to PATH without properly handling the Scripts directory.

Specifics:
- `python` command works normally
- `pip` command is not found
- But pip can be used via `python -m pip`

In comparison, pyenv correctly sets up the Python environment, including the Scripts directory, making pip available globally.

I hope the development team can address this issue to make mise's Python management functionality more complete.

**Reply by polarathene** [reply]
Pretty sure `pip` isn't required to be available in Python installations and that you can encounter this on some Linux distros as well. In those cases it's treated as a separate package to install, which is probably what should be handled with mise.

I'd recommend going with `uv`, which can acquire a specific version of Python for you (_same as mise provides, Python build standalone_).

**Comment by killerddd3** [comment]
Similar issue here.

**Comment by jylenhofdecat** [comment]
Using pip directly is a bad practice !
You should always use `python -m pip` and not just `pip`

See documentation :
https://pip.pypa.io/en/latest/user_guide/
https://docs.python.org/fr/dev/installing/index.html

Also, mise use python-build-standalone python builds by default if you don't enable compilation, and this is explained in documentation here for windows :
https://gregoryszorc.com/docs/python-build-standalone/main/quirks.html#no-pip-exe-on-windows

**Reply by jylenhofdecat** [reply]
And like other points, uv is gaining a lot of traction, you should use it ! It's fast

**Reply by DiefBell** [reply]
@jylenhofdecat Yes but not every third-party package follows this practice. So I need pip in my path, regardless of what I do

**Comment by JamBalaya56562** [comment]
`pip` and `pip3` are now fixed by #11278, which has been merged — it isn't in v2026.7.13, so it will ship in the next release. The `python3` half of the original report was fixed separately by #11212 and is already out in **v2026.7.13**.

**What was actually broken** — three separate things, which is why the workarounds in this thread only ever fixed part of it:

1. **`Scripts` was never on `PATH`.** mise only exposed the install root on Windows, but `pip.exe` and every console script installed by pip (`black.exe`, …) land in `<install>\Scripts`. That is exactly what @bharat-mon and @crazydan spotted — pip itself prints `WARNING: The scripts pip.exe, pip3.13.exe and pip3.exe are installed in '...\Scripts' which is not on PATH`, and manually adding that directory was the only way through. mise now includes `Scripts`, so it works without manual PATH edits, and shims are generated for what's in it.

2. **python-build-standalone ships no `pip.exe` at all on Windows.** @xvrpop's link was the right diagnosis. pip is preinstalled only as a module, so `python -m pip` works while `pip` does not exist as a command. Worse, the obvious fix doesn't help: `python -m pip install --upgrade pip` is a no-op while pip is current, which is why @gjlmotea's attempt printed `Requirement already satisfied` and no launcher ever appeared. mise now creates `pip.cmd`/`pip3.cmd` next to `python.exe` at install time, delegating to `python -m pip`. Because they delegate, they keep working after pip upgrades itself.

3. **Bonus:** `python.default_packages_file` silently failed on Windows because it spawned the unix-only `<install>/bin/python`. That was reported separately in #8872 and is fixed in the same PR.

**Two practical notes for existing setups:**

- The wrappers are created during installation, so an already-installed python won't have them. After upgrading mise, reinstall to pick them up: `mise install python@3.12 --force` (new installs get them automatically). The `Scripts` PATH entry applies immediately, no reinstall needed.
- If you use shims rather than `mise activate`, run `mise reshim` after a `pip install` so newly added console scripts get shims.

@jylenhofdecat is right that `python -m pip` is the most robust invocation, and it is unaffected — that is precisely what the wrappers call. The goal here is parity with mise on macOS/Linux, where `pip` is on `PATH` and the countless tutorials and CI scripts that assume a bare `pip` just work.

`docs/lang/python.md` now has a Windows section documenting all of this. Thanks everyone for the detailed reports — the pip `WARNING` output and the python-build-standalone quirks link made the root causes unambiguous.
